### ISSUE TITLE
Support for custom or shorter base workdir paths

### ISSUE THREAD

**Issue by MichaelJJ** [actions/runner#1676]
Currently, the base working directory for actions exists within a sub folder of the installation directory and duplicates the repository twice. This can be an issue for repos with longer names along with certain tools (Python VENV, MSBuild).

There is a workflow yaml option to set the base working directory, but this is not very portable and the path can be runner specific.

A few things could help with this issue:

1. Instead of using the pattern `<install_dir>/_work/<repo_name>/<repo_name>/` use the path `<install_dir>/_work/<guid_or_short_unique_string/`
2. Add a setting or environment variable to change the base path of `<install_dir>/_work/` to a different path on disk.

**Comment by MichaelJJ** [comment]
To add, https://github.com/actions/checkout also seems to only checkout to the default working directory, which is not possible on systems where long path support can't be enabled.

**Comment by ruvceskistefan** [comment]
Hi @MichaelJJ,
Thanks for the reported issue. I will label this issue as a _service feature_, so we will work on that in the coming period.

**Comment by MichaelJJ** [comment]
Great, thanks. This is causing some pain for multiple members of our Enterprise and would be great to have resolved!

**Comment by MichaelJJ** [comment]
Hello, any update on this?

**Comment by chrisnjohnson** [comment]
I am running into this issue as well and would also appreciate an update on it.

It would be great to have the ability to control the setting of the `GITHUB_WORKSPACE` variable so that the github checkout action doesn't require checking out the repo into such a long base path.  Currently the path is set to `<runner_install_dir>/_work/<repo_name>/<repo_name>`, and during the runner install I have the ability to control the `<runner_install_dir>/_work` location, but there is currently no way to control the `<repo_name>/<repo_name>` part of the path.  Being able to change this to just a single `<repo_name>` or something else that I can control would be extremely helpful.  Thanks!

**Comment by jsvasish** [comment]
I agree that many teams will benefit from this feature where long path names arent supported. We too ran into this issue with our python venv but we didnt receive any solution. We ended up creating our namespace & repo names with shorter names.

**Comment by kellycouch** [comment]
Visual Studio NMAKE is an example of popular tooling that has a 256 character path limit. We have a project repo is large and has deep, nested directories resulting in long paths. Combined with GitHub current work directory scheme our build paths are > 256 characters and results build errors.

Documented limitation:
https://docs.microsoft.com/en-us/cpp/error-messages/tool-errors/nmake-fatal-error-u1076?view=msvc-170

**Comment by baparham** [comment]
@ruvceskistefan any updates on this?

**Comment by fhammerl** [comment]
For users struggling with this on self-hosted runners:

> Add a setting or environment variable to change the base path of <install_dir>/_work/ to a different path on disk.

As you run `config.sh` configure a runner, you have an option to set the working directory. Run `config.sh` manually or use `config.sh <other_params> --work /target/dir/` to try this.

Note that `_work` doesn’t have to be under <install_dir> - I hope this can provide some relief.

**Comment by rentziass** [comment]
As a workaround to this you could also create a symlink to `$GITHUB_WORKSPACE` to shorten the path. For example in an Ubuntu runner you could

```yaml
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Symlink
        run: ln -s $GITHUB_WORKSPACE /home/runner/work/project
```

**Comment by baparham** [comment]
Thanks for the tips! Setting the base work directory is helpful but only gets us so far. If your repo name is rather long (e.g. in an enterprise environment where repo names may have fixed hierarchical names and be quite long) the fact that we can't rename the folder the runner puts the repo in _underneath_ the work folder is the root problem with long paths here.

**Comment by kraszkow** [comment]
> Hi @MichaelJJ, Thanks for the reported issue. I will label this issue as a _runner feature_, so we will work on that in the coming period.

@ruvceskistefan - any update on this? 


> For users struggling with this on self-hosted runners:
> 
> > Add a setting or environment variable to change the base path of <install_dir>/_work/ to a different path on disk.
> 
> As you run `config.sh` configure a runner, you have an option to set the working directory. Run `config.sh` manually or use `config.sh <other_params> --work /target/dir/` to try this.
> 
> Note that `_work` doesn’t have to be under <install_dir> - I hope this can provide some relief.

@fhammerl - unfortunately it doesn't help. Our biggest problem is that repo_name is used 2 times: `/<repo_name>/<repo_name>/`

**Comment by ReneSorensenBaader** [comment]
This is a serious matter, i have 60+ repos that does not build because of this issue. Mainly this comes down to msbuild do not support long path names. 
The fix @kraszkow metions, is what i am trying to implement at the moment, please make something happen...

**Comment by MichaelJJ** [comment]
Hello, any updates on this request? Would love to see some traction on this request...

**Comment by StephenHodgson** [comment]
I'm trying to figure out why the repo name is in there twice 🤔

**Comment by StephenHodgson** [comment]
I was just looking for that! thanks

**Comment by tbranch227** [comment]
> 3\. Repositories

With 100s of repositories and using self-hosted runners, we run into the same issues.  We actually maintain our old working folders for a few days, so that we can review tasks later for why things failed.  Sometimes it takes a while for a  pattern to emerge that let's us know we need to review past builds.  

When you are building mobile and desktop apps, the process gets complicated quickly and you're upgrading tools every quarter to keep up with your target platforms.  

Setting that folder at runtime.  I thought the hook scripts might give me that opportunity to set a working folder that would be used by the action, but the environment variable is overwritten by the action or the action runner.  I'm not sure, but the environment variable definitely gets reset.

We need to be able to do the following:
  * set our working folders at runtime and have that respected by 3rd party actions
  * be able to set these parameters with the hook scripts rather than the workflow files to maintain standardization across the enterprise and reduce maintenance.  We will have 100s of workflows and tracking that everyone is using the right folder in their workflow files, simply isn't maintainable.

Thanks - for now - looks like we'll be using all custom actions.

**Comment by jahales-intel** [comment]
Someone from my organization pointed me to a workaround that they're using and I'm sharing here in case it helps while we wait for a robust fix. 

There's a Pipeline configuration folder in the runner work directory where you can modify the pipeline directory and workspace directory for a given repository:

<work directory>\PipelineMapping\<organization>\<repository>\PipelineFolder.json

I've been modifying that file to create a shorter file path and it seems to work (pipelineDirectory / workspaceDirectory). I'm not aware of any documentation for this.

**Comment by skharel146** [comment]
@jahales-intel its going to be tedious if you have a lot of repositories to create at once, isn't it?

**Comment by jahales-intel** [comment]
> @jahales-intel its going to be tedious if you have a lot of repositories to create at once, isn't it?

This certainly isn't a great long term solution.

**Comment by sergii-rybin-tfs** [comment]
Hello, any updates on this request? We have the same problem with many repositories.

**Comment by skharel146** [comment]
Same is the case for me, I wrote a PS script to read the PipelineMappings.json file and update the values.

**Comment by interifter** [comment]
Anyone know how to get attention on this issue?

**Comment by Piedone** [comment]
Some more context: https://github.com/actions/checkout/issues/955.

**Comment by Danielku15** [comment]
Same problem here. I'm migrating many repositories in our GitHub enterprise instance from TeamCity to GitHub Actions and boom: Max Path exceeded for various repositories. 

Form the discussion I understand its important to have separate folders.

It might be a bit short-sighted but why not simply do something like this: 

https://github.com/actions/runner/blob/04b07b6675c56da40e532ab73eeb4287bf223e34/src/Runner.Worker/TrackingConfig.cs#L60-L62
```diff
            // Set the directories.
            PipelineDirectory = repoName.ToString(CultureInfo.InvariantCulture);
-            WorkspaceDirectory = Path.Combine(PipelineDirectory, repoName);
+            WorkspaceDirectory = Path.Combine(PipelineDirectory, "w"); // short sub-path for workspace
```

This still eats 2 characters from the path but its definitly better than now.

**Comment by milos-licina-3shape** [comment]
We are also experiencing the same issue. The way we name the repositories and the way we structure the code coupled with the double nesting from GitHub leads to a long default root directory of our code. When we try to compile our code and run tests afterwards we are hitting limits with certain parts of the dotnet tooling not respecting long path registry settings in Windows.

We would really appreciate a supported and documented way of changing the Workspace directory, without reverse engineering GitHub logic. If you are worried on the impact of this on GitHub hosted runners, a self hosted only feature would be greatly appreciated.

**Comment by masa-ibara** [comment]
Any progress on this issue?
It's unfortunate that we have to give up using GitHub Actions because of this.

**Comment by duaneking** [comment]
Just wanted to how my support for this fix as its actively blocking work I want to do; A self hosted runner is not an option due to security compliance requirements.

I can enable git's long path support but that doesn't help me because of windows not respecting registry settings in Net/VS tooling

**Comment by u011** [comment]
Feature is such a mislabel, it's related to a big bug!

and it fails on git submodule init (lol github) and on ai development for open source (lol msft)

build / win32-x64
unable to write file D:/a/my-very-longy-long-repo-name-here/my-very-longy-long-repo-name-here/.git/modules/addon/vendor/mlc/mlc-llm/modules/3rdparty/tvm/modules/3rdparty/flashinfer/modules/3rdparty/composable_kernels/objects/pack/pack-146ca6f59817db32764fbf97e6ac6be89111c5ad.pack: No such file or directory

**Comment by keremispirli** [comment]
If GitHub claims Windows compatibility for GitHub Runners then this is indeed a BUG, not a feature.
